### ISSUE TITLE
[1d] IMPORTANT: dm push sometimes hangs

### ISSUE THREAD

**Issue by lukemarsden** [dotmesh-io/dotmesh#586]
a transferRequest
```
"interclusterTransfers.5536f85f-5b19-4ba8-7f76-2b6d19b0252d": "{\"TransferRequestId\":\"5536f85f-5b19-4ba8-7f76-2b6d19b0252d\",\"Peer\":\"cloud.dotscience.net\",\"User\":\"lukedemo49\",\"ApiKey\":\"XXX====\",\"Direction\":\"pull\",\"LocalNamespace\":\"admin\",\"LocalName\":\"combined-houses\",\"LocalBranchName\":\"\",\"RemoteNamespace\":\"lukedemo49\",\"RemoteName\":\"combined-houses\",\"RemoteBranchName\":\"\",\"FilesystemId\":\"20d3de41-3a9e-4bab-5de2-8c97b98ff5d9\",\"InitiatorNodeId\":\"ab54ac77a273e3b1\",\"PeerNodeId\":\"\",\"StartingCommit\":\"START\",\"TargetCommit\":\"b25c65fe-ff44-4412-5456-e904cff447b5\",\"Index\":1,\"Total\":1,\"Status\":\"finished\",\"NanosecondsElapsed\":72639681,\"Size\":12912,\"Sent\":49807,\"Message\":\"\"}",```
```

seems to make `dm` hang
```
Pushing admin/project-e219e493-default-workspace to hub:lukedemo49/project-e219e493-default-workspace
Calculating...
```

stacks show it was sleeping at the time
```
Pushing admin/project-e219e493-default-workspace to hub:lukedemo49/project-e219e493-default-workspace
Calculating...
SIGQUIT: quit
PC=0x459a31 m=0 sigcode=0
goroutine 0 [idle]:
runtime.futex(0x12656a8, 0x0, 0x0, 0x0, 0x0, 0x126d560, 0x0, 0x0, 0x7ffc5ad5b2d8, 0x40fe1b, ...)
        bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_static_stripped/stdlib~/src/runtime/sys_linux_amd64.s:530 +0x21
runtime.futexsleep(0x12656a8, 0x7efc00000000, 0xffffffffffffffff)
        GOROOT/src/runtime/os_linux.go:45 +0x4b
runtime.notesleep(0x12656a8)
        GOROOT/src/runtime/lock_futex.go:151 +0x9b
runtime.stopm()
        GOROOT/src/runtime/proc.go:1952 +0xe5
runtime.findrunnable(0xc420032500, 0x0)
        GOROOT/src/runtime/proc.go:2415 +0x50c
runtime.schedule()
        GOROOT/src/runtime/proc.go:2541 +0x13b
runtime.park_m(0xc420001080)
        GOROOT/src/runtime/proc.go:2604 +0xb6
runtime.mcall(0xab74d0)
        bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_static_stripped/stdlib~/src/runtime/asm_amd64.s:351 +0x5b
goroutine 1 [sleep]:
time.Sleep(0x3b9aca00)
        GOROOT/src/runtime/time.go:102 +0x166
github.com/dotmesh-io/dotmesh/cmd/dm/pkg/remotes.(*DotmeshAPI).PollTransfer(0xc42021e570, 0xc42026e060, 0x24, 0xe24840, 0xc42000e018, 0x0, 0x0)
        cmd/dm/pkg/remotes/api.go:748 +0x10f
github.com/dotmesh-io/dotmesh/cmd/dm/pkg/commands.NewCmdPush.func1.1(0xc420071c80, 0x2, 0x4, 0xe24840, 0xc42000e018, 0xd65780, 0xc420020a60)
        cmd/dm/pkg/commands/push.go:52 +0x162
github.com/dotmesh-io/dotmesh/cmd/dm/pkg/commands.NewCmdPush.func1(0xc42021cd80, 0xc420071c80, 0x2, 0x4)
        cmd/dm/pkg/commands/push.go:57 +0x5b
github.com/dotmesh-io/dotmesh/cmd/dm/vendor/github.com/spf13/cobra.(*Command).execute(0xc42021cd80, 0xc420071c40, 0x4, 0x4, 0xc42021cd80, 0xc420071c40)
        cmd/dm/vendor/github.com/spf13/cobra/command.go:636 +0x234
github.com/dotmesh-io/dotmesh/cmd/dm/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x122a4c0, 0xc420215d40, 0xc420215b00, 0xc4200a9680)
        cmd/dm/vendor/github.com/spf13/cobra/command.go:722 +0x2d4
github.com/dotmesh-io/dotmesh/cmd/dm/vendor/github.com/spf13/cobra.(*Command).Execute(0x122a4c0, 0xc4201017a0, 0xc420209f58)
        cmd/dm/vendor/github.com/spf13/cobra/command.go:681 +0x2b
main.main()
        cmd/dm/dotmesh.go:40 +0x1f9
goroutine 5 [syscall, 2 minutes]:
os/signal.signal_recv(0x0)
        GOROOT/src/runtime/sigqueue.go:139 +0xa6
os/signal.loop()
        GOROOT/src/os/signal/signal_unix.go:22 +0x22
created by os/signal.init.0
        GOROOT/src/os/signal/signal_unix.go:28 +0x41
goroutine 13 [select]:
github.com/dotmesh-io/dotmesh/cmd/dm/vendor/gopkg.in/cheggaaa/pb%2ev1.(*ProgressBar).refresher(0xc42009da40)
        cmd/dm/vendor/gopkg.in/cheggaaa/pb.v1/pb.go:430 +0xf2
created by github.com/dotmesh-io/dotmesh/cmd/dm/vendor/gopkg.in/cheggaaa/pb%2ev1.(*ProgressBar).Start
        cmd/dm/vendor/gopkg.in/cheggaaa/pb.v1/pb.go:123 +0xc0
goroutine 10 [IO wait]:
internal/poll.runtime_pollWait(0x7efcf0846f00, 0x72, 0xc4200509a8)
        GOROOT/src/runtime/netpoll.go:173 +0x57
internal/poll.(*pollDesc).wait(0xc42015c318, 0x72, 0xffffffffffffff00, 0xe25720, 0x1232d80)
        GOROOT/src/internal/poll/fd_poll_runtime.go:85 +0x9b
internal/poll.(*pollDesc).waitRead(0xc42015c318, 0xc42021b000, 0x1000, 0x1000)
        GOROOT/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Read(0xc42015c300, 0xc42021b000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
        GOROOT/src/internal/poll/fd_unix.go:157 +0x17d
net.(*netFD).Read(0xc42015c300, 0xc42021b000, 0x1000, 0x1000, 0x454390, 0xc420151980, 0x4)
        GOROOT/src/net/fd_unix.go:202 +0x4f
net.(*conn).Read(0xc42000e868, 0xc42021b000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
        GOROOT/src/net/net.go:176 +0x6a
net/http.(*persistConn).Read(0xc420230240, 0xc42021b000, 0x1000, 0x1000, 0xc420050b98, 0x404435, 0xc420238000)
        GOROOT/src/net/http/transport.go:1453 +0x136
bufio.(*Reader).fill(0xc42001f740)
        GOROOT/src/bufio/bufio.go:100 +0x11e
bufio.(*Reader).Peek(0xc42001f740, 0x1, 0x0, 0x0, 0x0, 0xc420238d20, 0x0)
        GOROOT/src/bufio/bufio.go:132 +0x3a
net/http.(*persistConn).readLoop(0xc420230240)
        GOROOT/src/net/http/transport.go:1601 +0x185
created by net/http.(*Transport).dialConn
        GOROOT/src/net/http/transport.go:1237 +0x95a
goroutine 11 [select]:
net/http.(*persistConn).writeLoop(0xc420230240)
        GOROOT/src/net/http/transport.go:1822 +0x14b
created by net/http.(*Transport).dialConn
        GOROOT/src/net/http/transport.go:1238 +0x97f
rax    0xca
rbx    0x1265560
rcx    0x459a33
rdx    0x0
rdi    0x12656a8
rsi    0x0
rbp    0x7ffc5ad5b2a0
rsp    0x7ffc5ad5b258
r8     0x0
r9     0x0
```

how the heck can `func (dm *DotmeshAPI) PollTransfer` fail to return in that case?

versions
```
root@dotscience-runner-luke-demo49b:~# docker exec -ti dotscience-committer-cf66750a bash
root@88b0b9302970:/usr/src/app# dm version
Current remote: local (use 'dm remote -v' to list and 'dm remote switch' to switch)
Client:
        Version: master-7374592
Server:
        Version: master-faa4111
```

**Comment by lukemarsden** [comment]
eek, saw this
```
  "interclusterTransfers.c11aa5b9-9254-412b-4102-a4ed88d676cc": "{\"TransferRequestId\":\"c11aa5b9-9254-412b-4102-a4ed88d676cc\",\"Peer\":\"cloud.dotscience.net\",\"User\":\"lukedemo49\",\"ApiKey\":\"XXX====\",\"Direction\":\"push\",\"LocalNamespace\":\"admin\",\"LocalName\":\"project-e219e493-default-workspace\",\"LocalBranchName\":\"\",\"RemoteNamespace\":\"lukedemo49\",\"RemoteName\":\"project-e219e493-default-workspace\",\"RemoteBranchName\":\"\",\"FilesystemId\":\"0437c0d8-678f-4dc1-52b7-ac2a752cd4d0\",\"InitiatorNodeId\":\"ab54ac77a273e3b1\",\"PeerNodeId\":\"\",\"StartingCommit\":\"d570ef8f-569c-4dfe-4729-2bc0375d18bf\",\"TargetCommit\":\"cf9de5cd-6d05-4d68-62ce-6751db99f80b\",\"Index\":1,\"Total\":1,\"Status\":\"pushing\",\"NanosecondsElapsed\":1244939259,\"Size\":150824,\"Sent\":216755,\"Message\":\"\"}",
```

**Comment by alaric-dotmesh** [comment]
The changes I've committed will hopefully help.

We looked into how the push was ending up hanging. Logging revealed that it *was* set to the `finished` state then set back to `pushing` again a moment later, leaving the client waiting forever for it to reach `finished`.

The code responsible in the server updated a fsMachine-global poll result object all over the place (including in different goroutines) and called `updatePollResult` to send it to etcd.

It was very hard to follow who updated what fields of it last, and the fact that multiple goroutines were involved (without locking) meant we were also drifting into "undefined behaviour" territory, so the changes committed above create a new goroutine per fsmachine that's responsible for all mutations to the poll result, fed by messages sent to it over a channel.

This means that all updates to the poll result are logged under `updateEtcdAboutTransfers` so can be easily found and traced, and that we're not producing code with undefined behaviour.

This will either fix the problem outright, or make it easier to trace down what's causing it if it recurs...
